sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeSet, ExpressionSet, NamedExpression}
+import org.apache.spark.sql.catalyst.plans.LeftExistence
+
+/**
+ * A visitor pattern for traversing a [[LogicalPlan]] tree and propagate the distinct attributes.
+ */
+object DistinctKeyVisitor extends LogicalPlanVisitor[Set[AttributeSet]] {
+
+  private def projectDistinctKeys(
+      keys: Set[ExpressionSet], projectList: Seq[NamedExpression]): Set[AttributeSet] = {
+    val expressions = keys.flatMap(_.toSet)
+    projectList.filter {
+      case a: Alias => expressions.exists(_.semanticEquals(a.child))
+      case ne => expressions.exists(_.semanticEquals(ne))
+    }.toSet.subsets(keys.map(_.size).min).filter { s =>
+      val references = s.map {
+        case a: Alias => a.child
+        case ne => ne
+      }
+      keys.exists(_.equals(ExpressionSet(references)))
+    }.map(s => AttributeSet(s.map(_.toAttribute))).toSet
```

Comment:
how about
```
    val outputSet = ExpressionSet(projectList.map(_.toAttribute))
    val aliases = projectList.filter(_.isInstanceOf[Alias])
    if (aliases.isEmpty) return keys.filter(_.subsetOf(outputSet))

    val aliasedDistinctKeys = keys.map { expressionSet =>
      expressionSet.map { expression =>
        expression transform {
          case expr: Expression =>
            aliases
              .collectFirst { case a: Alias if a.child.semanticEquals(expr) => a.toAttribute }
              .getOrElse(expr)
        }
      }
    }
    aliasedDistinctKeys.collect {
      case es: ExpressionSet if es.subsetOf(outputSet) => ExpressionSet(es)
    }
```

Reply:
If one expression has multiple aliases, we need to further expand the distinct keys set. We can do it later as it's rather a corner case.

Reply:
OK